# Ex 4.8 - Grouping Data & Aggregating Variables

### This script contains the following points:

#### Task 2. Grouping Data.
#### Task 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset?
#### Task 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.
#### Task 5. Determine whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.
#### Task 6. Create a spending flag for each user.
#### Task 7. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column.
#### Task 09. Export your dataframe as a pickle file and store it correctly in your 'Prepared Data' folder.

In [130]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [131]:
# Turn project folder path into a string.
path = '/Users/asadagha/Documents/CF Data Analytics/Data Analytics Immersion/Achievement 4/Instacart Basket Analysis'

In [132]:
# Import the dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_derived.pkl'))

### Task 2. Grouping Data.

In [134]:
# Grouping by 'department_id' and 'order_number', and Calculating Mean using agg() Function
ords_prods_merge.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Task 3. Analyze the result. How do the results for the entire dataframe differ from those of the subset?

The mean values for the department ids present in the subset are different from the that of the full dataset, some are higher while others are lower than that of the entire dataset. Furthermore, the subset didn't have records with all the deoartment ids.


### Task 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [138]:
# Aggregating Data with transform() Function
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

/var/folders/43/fcmpd7jx0z18k499r5yr6q7r0000gn/T/ipykernel_13597/1549965947.py:2: FutureWarning: The provided callable <function max at 0x11072fec0> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [139]:
# Checking the Output
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    3139998      138            28                  6                 11   
1    1977647      138            30                  6                 17   
2     389851      709             2                  0                 21   
3     652770      764             1                  3                 13   
4    1813452      764             3                  4                 17   
5    1701441      777            16                  1                  7   
6    1871483      825             3                  2                 14   
7    1290456      910            12                  3                 10   
8     369558     1052            10                  1                 20   
9     589712     1052            15                  1                 12   
10   1798115     1379             1                  5                 11   
11    497193     1480            36                  3                 13   
12   1727826     1480            38                  4                 13   
13   2694698     1480            39                  2                 12   
14   3171753     1494             6                  3                 10   

    ...  first_order  add_to_cart_order  reordered  _merge merged_dataframes  \
0   ...        False                  5          0    both              both   
1   ...        False                  1          1    both              both   
2   ...        False                 20          0    both              both   
3   ...         True                 10          0    both              both   
4   ...        False                 11          1    both              both   
5   ...        False                  7          0    both              both   
6   ...        False                  2          0    both              both   
7   ...        False                  1          0    both              both   
8   ...        False                  1          0    both              both   
9   ...        False                  2          1    both              both   
10  ...         True                  1          0    both              both   
11  ...        False                  1          0    both              both   
12  ...        False                  3          1    both              both   
13  ...        False                  4          1    both              both   
14  ...        False                  8          0    both              both   

      price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regularly busy           Most orders   
1   Mid-range product  Regularly busy  Regularly busy        Averag

In [140]:
# Create 'Loyalty' Flags using loc Function
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [141]:
# Check Value Counts of 'loyalty_flag'
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [142]:
# Checking the Updated Output
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    3139998      138            28                  6                 11   
1    1977647      138            30                  6                 17   
2     389851      709             2                  0                 21   
3     652770      764             1                  3                 13   
4    1813452      764             3                  4                 17   
5    1701441      777            16                  1                  7   
6    1871483      825             3                  2                 14   
7    1290456      910            12                  3                 10   
8     369558     1052            10                  1                 20   
9     589712     1052            15                  1                 12   
10   1798115     1379             1                  5                 11   
11    497193     1480            36                  3                 13   
12   1727826     1480            38                  4                 13   
13   2694698     1480            39                  2                 12   
14   3171753     1494             6                  3                 10   

    ...  add_to_cart_order  reordered  _merge  merged_dataframes  \
0   ...                  5          0    both               both   
1   ...                  1          1    both               both   
2   ...                 20          0    both               both   
3   ...                 10          0    both               both   
4   ...                 11          1    both               both   
5   ...                  7          0    both               both   
6   ...                  2          0    both               both   
7   ...                  1          0    both               both   
8   ...                  1          0    both               both   
9   ...                  2          1    both               both   
10  ...                  1          0    both               both   
11  ...                  1          0    both               both   
12  ...                  3          1    both               both   
13  ...                  4          1    both               both   
14  ...                  8          0    both               both   

      price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regularly busy           Most orders   
1   Mid-range product  Regularly busy  Regularly busy        Average orders   
2   Mid-range product     Busiest day    Busiest days        Average orders   
3   Mid-range product  Regularly busy    Slowest days           Most orders   
4   Mid-range product 

### Task 5. Determine whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [144]:
# Determine basic statistics of customer spending habits based on 'loyalty_flag' categories
ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Those amongst the 'Loyal customer' category purchase items with a mean price of 10.39. Whereas those amongst the 'New customer' and 'Regular customer' categories both purchase items with a mean price of 13.29 and 12.50 respectively, both higher than that of items purchased by loyal customers.

### Task 6. Create a spending flag for each user based on the average price across all their orders using the following criteria:
1. If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender”
2. If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [147]:
# Create a new column for the average spend based on the price of the products purchased by a customer
ords_prods_merge['average_spend'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [148]:
# Check the new column
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    3139998      138            28                  6                 11   
1    1977647      138            30                  6                 17   
2     389851      709             2                  0                 21   
3     652770      764             1                  3                 13   
4    1813452      764             3                  4                 17   
5    1701441      777            16                  1                  7   
6    1871483      825             3                  2                 14   
7    1290456      910            12                  3                 10   
8     369558     1052            10                  1                 20   
9     589712     1052            15                  1                 12   
10   1798115     1379             1                  5                 11   
11    497193     1480            36                  3                 13   
12   1727826     1480            38                  4                 13   
13   2694698     1480            39                  2                 12   
14   3171753     1494             6                  3                 10   

    ...  reordered  _merge  merged_dataframes    price_range_loc  \
0   ...          0    both               both  Mid-range product   
1   ...          1    both               both  Mid-range product   
2   ...          0    both               both  Mid-range product   
3   ...          0    both               both  Mid-range product   
4   ...          1    both               both  Mid-range product   
5   ...          0    both               both  Mid-range product   
6   ...          0    both               both  Mid-range product   
7   ...          0    both               both  Mid-range product   
8   ...          0    both               both  Mid-range product   
9   ...          1    both               both  Mid-range product   
10  ...          0    both               both  Mid-range product   
11  ...          0    both               both  Mid-range product   
12  ...          1    both               both  Mid-range product   
13  ...          1    both               both  Mid-range product   
14  ...          0    both               both  Mid-range product   

       busiest_day    busiest_days busiest_period_of_day max_order  \
0   Regularly busy  Regularly busy           Most orders        32   
1   Regularly busy  Regularly busy        Average orders        32   
2      Busiest day    Busiest days        Average orders         5   
3   Regularly busy    Slowest days           Most orders         3   
4       Least busy    Slowest days        Average orders         3 

In [149]:
# Assign either 'Low_spender' or 'High_spender' tag to each customer
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'
ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [150]:
# Check value counts of spender flags
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     31770614
High_spender      634245
Name: count, dtype: int64

In [151]:
# Check Releavent Columns
ords_prods_merge[['user_id', 'average_spend', 'spender_flag']].head(15)

user_id  average_spend spender_flag
0       138       6.935811  Low_spender
1       138       6.935811  Low_spender
2       709       7.930208  Low_spender
3       764       4.972414  Low_spender
4       764       4.972414  Low_spender
5       777       6.935398  Low_spender
6       825       5.957576  Low_spender
7       910       6.680000  Low_spender
8      1052       7.162500  Low_spender
9      1052       7.162500  Low_spender
10     1379       5.813636  Low_spender
11     1480       5.081633  Low_spender
12     1480       5.081633  Low_spender
13     1480       5.081633  Low_spender
14     1494       6.803797  Low_spender

### Task 7. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
1. If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a 'Non-frequent customer'
2. If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a 'Regular customer'
3. If the median is lower than or equal to 10, then the customer should be labeled a 'Frequent customer'

In [153]:
# Create a new column for customer frequency flags based on the median values from 'days_since_prior_order' column
ords_prods_merge['customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [154]:
# Check the newly created column
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    3139998      138            28                  6                 11   
1    1977647      138            30                  6                 17   
2     389851      709             2                  0                 21   
3     652770      764             1                  3                 13   
4    1813452      764             3                  4                 17   
5    1701441      777            16                  1                  7   
6    1871483      825             3                  2                 14   
7    1290456      910            12                  3                 10   
8     369558     1052            10                  1                 20   
9     589712     1052            15                  1                 12   
10   1798115     1379             1                  5                 11   
11    497193     1480            36                  3                 13   
12   1727826     1480            38                  4                 13   
13   2694698     1480            39                  2                 12   
14   3171753     1494             6                  3                 10   

    ...  merged_dataframes    price_range_loc     busiest_day    busiest_days  \
0   ...               both  Mid-range product  Regularly busy  Regularly busy   
1   ...               both  Mid-range product  Regularly busy  Regularly busy   
2   ...               both  Mid-range product     Busiest day    Busiest days   
3   ...               both  Mid-range product  Regularly busy    Slowest days   
4   ...               both  Mid-range product      Least busy    Slowest days   
5   ...               both  Mid-range product  Regularly busy    Busiest days   
6   ...               both  Mid-range product  Regularly busy  Regularly busy   
7   ...               both  Mid-range product  Regularly busy    Slowest days   
8   ...               both  Mid-range product  Regularly busy    Busiest days   
9   ...               both  Mid-range product  Regularly busy    Busiest days   
10  ...               both  Mid-range product  Regularly busy  Regularly busy   
11  ...               both  Mid-range product  Regularly busy    Slowest days   
12  ...               both  Mid-range product      Least busy    Slowest days   
13  ...               both  Mid-range product  Regularly busy  Regularly busy   
14  ...               both  Mid-range product  Regularly busy    Slowest days   

   busiest_period_of_day max_order      loyalty_flag average_spend  \
0            Most orders        32  Regular customer      6.935811   
1         Average orders        32  Regular customer      6.935811   

In [155]:
# Based on the criteria provided, create a column featuring order frequency flags
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] > 20, 'frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['customer_frequency'] > 10) & (ords_prods_merge['customer_frequency'] <=20), 'frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [156]:
# Check value counts for new column
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [157]:
# Check columns of updated dataframe
ords_prods_merge[['user_id', 'customer_frequency', 'frequency_flag']].head(20)

user_id  customer_frequency     frequency_flag
0       138                 8.0  Frequent customer
1       138                 8.0  Frequent customer
2       709                 8.0  Frequent customer
3       764                 9.0  Frequent customer
4       764                 9.0  Frequent customer
5       777                11.0   Regular customer
6       825                20.0   Regular customer
7       910                 6.0  Frequent customer
8      1052                10.0  Frequent customer
9      1052                10.0  Frequent customer
10     1379                20.0   Regular customer
11     1480                 7.0  Frequent customer
12     1480                 7.0  Frequent customer
13     1480                 7.0  Frequent customer
14     1494                 7.0  Frequent customer
15     1494                 7.0  Frequent customer
16     1494                 7.0  Frequent customer
17     1540                 7.0  Frequent customer
18     1540                 7.0  Frequent customer
19     1540                 7.0  Frequent customer

### Task 09. Export your dataframe as a pickle file and store it correctly in your 'Prepared Data' folder.

In [159]:
# Check shape before exporting
ords_prods_merge.shape

(32404859, 26)

In [160]:
# Export the updated dataframe as a pickle file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged_grouped.pkl'))